core/src/services/vercel_blob/core.rs

```diff
@@ -0,0 +1,423 @@
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+use std::fmt::Debug;
+use std::fmt::Formatter;
+
+use bytes::Bytes;
+use http::header;
+use http::request;
+use http::Request;
+use http::Response;
+use http::StatusCode;
+use serde::Deserialize;
+use serde::Serialize;
+use serde_json::json;
+
+use crate::raw::*;
+use crate::*;
+
+use super::error::parse_error;
+
+#[derive(Clone)]
+pub struct VercelBlobCore {
+    /// The root of this core.
+    pub root: String,
+    /// Vercel Blob token.
+    pub token: String,
+
+    pub client: HttpClient,
+}
+
+impl Debug for VercelBlobCore {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("Backend")
+            .field("root", &self.root)
+            .finish_non_exhaustive()
+    }
+}
+
+impl VercelBlobCore {
+    #[inline]
+    pub async fn send(&self, req: Request<AsyncBody>) -> Result<Response<IncomingAsyncBody>> {
+        self.client.send(req).await
+    }
+
+    pub fn sign(&self, req: request::Builder) -> request::Builder {
+        req.header(header::AUTHORIZATION, format!("Bearer {}", self.token))
+    }
+}
+
+impl VercelBlobCore {
+    pub async fn download(&self, path: &str, args: OpRead) -> Result<Response<IncomingAsyncBody>> {
+        let p = build_abs_path(&self.root, path);
+
+        let resp = self.list(&p, Some(1)).await?;
+
+        let blobs = resp.blobs;
+
+        if blobs.is_empty() {
+            return Err(Error::new(ErrorKind::NotFound, "Blob not found"));
+        }
+
+        let url = &blobs[0].url;
+
+        let mut req = Request::get(url);
+
+        let range = args.range();
+        if !range.is_full() {
+            req = req.header(http::header::RANGE, range.to_header());
+        }
+
+        // Set body
+        let req = req
+            .body(AsyncBody::Empty)
+            .map_err(new_request_build_error)?;
+
+        self.send(req).await
+    }
+
+    pub async fn get_put_request(
+        &self,
+        path: &str,
+        size: Option<u64>,
+        args: &OpWrite,
+        body: AsyncBody,
+    ) -> Result<Request<AsyncBody>> {
+        let p = build_abs_path(&self.root, path);
+
+        let url = format!(
+            "https://blob.vercel-storage.com/{}",
+            percent_encode_path(&p)
+        );
+
+        let mut req = Request::put(&url);
+
+        req = req.header("x-add-random-suffix", "0");
+
+        if let Some(size) = size {
+            req = req.header(header::CONTENT_LENGTH, size.to_string())
+        }
+
+        if let Some(mime) = args.content_type() {
+            req = req.header("x-content-type", mime)
```

Comment:
Please provide a reference for the specific behavior of the service so that we understand why it is necessary to write in this manner.

Reply:
fixed